### PR TITLE
Handshake

### PR DESCRIPTION
**Pull request by haaawk**

### CODE REVIEW

File: sqld/src/rpc/mod.rs [hunk 1]
```diff
@@ -21,14 +23,15 @@ pub async fn run_rpc_server<F>(
     ca_cert_path: Option<PathBuf>,
     factory: F,
     logger: Arc<WalLogger>,
+    hello_response: HelloResponse,
```
Comment (MarinPostma) [inline sqld/src/rpc/mod.rs:26]: I think that this information should be retrieved from the `WalLogger`

Comment (haaawk) [reply sqld/src/rpc/mod.rs:26]: done.

File: sqld/src/lib.rs [hunk 2]
```diff
@@ -167,7 +172,30 @@ pub async fn run_server(config: Config) -> anyhow::Result<()> {
...
+                Err(e) if e.kind() == ErrorKind::UnexpectedEof => {
+                    // OK: the file was empty, this is first generation
+                    0
+                }
+                Err(e) => Err(e)?,
+            };
+            generation_file
+                .write_all_at(&generation_id.to_le_bytes(), 0)
+                .unwrap();
+
```
[the remark this answers is not in the document]
Comment (MarinPostma) [reply sqld/src/lib.rs:196]: I think `WalLogger` should have a `generation` method that returns  the generation and the start frame index

File: sqld/src/rpc/wal_log.rs [hunk 3]
```diff
@@ -3,18 +3,25 @@ pub mod wal_log_rpc {
...
 
-use self::wal_log_rpc::{wal_log_entry::Payload, Frame, LogOffset, WalLogEntry as RpcWalLogEntry};
+use self::wal_log_rpc::{
+    wal_log_entry::Payload, Frame, HelloRequest, HelloResponse, LogOffset,
+    WalLogEntry as RpcWalLogEntry,
+};
 
 pub struct WalLogService {
     logger: Arc<WalLogger>,
+    replicas_with_hello: RwLock<Vec<SocketAddr>>,
```
Comment (MarinPostma) [inline sqld/src/rpc/wal_log.rs:23]: what about an `HashSet` here?

File: sqld/src/rpc/wal_log.rs [hunk 4]
```diff
@@ -74,8 +85,32 @@ impl WalLog for WalLogService {
...
+        req: tonic::Request<HelloRequest>,
+    ) -> Result<tonic::Response<HelloResponse>, Status> {
+        let replica_addr = req
+            .remote_addr()
+            .ok_or(Status::internal("No remote RPC address"))?;
+        {
+            let mut guard = self.replicas_with_hello.write().await;
+            if !guard.contains(&replica_addr) {
+                guard.push(replica_addr);
+            }
```
Comment (MarinPostma) [inline sqld/src/rpc/wal_log.rs:112]: let's just use an hashset to make this operation idempotent

File: sqld/src/rpc/wal_log.rs [hunk 5]
```diff
@@ -74,8 +85,32 @@ impl WalLog for WalLogService {
         &self,
         req: tonic::Request<LogOffset>,
     ) -> Result<tonic::Response<Self::LogEntriesStream>, Status> {
+        let replica_addr = req
+            .remote_addr()
+            .ok_or(Status::internal("No remote RPC address"))?;
+        {
+            let guard = self.replicas_with_hello.read().await;
+            if !guard.contains(&replica_addr) {
+                return Err(Status::failed_precondition("Hello message missing"));
```
Comment (MarinPostma) [inline sqld/src/rpc/wal_log.rs:94]: can we use something that look more like an error code, like `"NO_HELLO"`, and put that in a `const` so we can reuse that to check for the code on the client side

**Review by MarinPostma — CHANGES_REQUESTED** [review]

File: sqld/src/rpc/wal_log.rs (hunk 3 above)
Comment (haaawk) [reply sqld/src/rpc/wal_log.rs:23]: HashSet will perform worse unless we have thousands of replicas. Both cpu wise and memory wise.

File: sqld/src/rpc/wal_log.rs (hunk 4 above)
Comment (haaawk) [reply sqld/src/rpc/wal_log.rs:112]: hashset is an overkill when we expect << 100 replicas

Comment (MarinPostma) [reply sqld/src/rpc/wal_log.rs:112]: it's just more ergonomic, and expresses the intent better

File: sqld/src/rpc/wal_log.rs (hunk 3 above)
Comment (MarinPostma) [reply sqld/src/rpc/wal_log.rs:23]: "early optimization is the root of all evil" 😄 This is clearly not on the hot path

File: sqld/src/rpc/wal_log.rs [hunk 6]
```diff
@@ -42,7 +52,15 @@ impl From<(u64, WalLogEntry)> for RpcWalLogEntry {
 
 impl WalLogService {
     pub fn new(logger: Arc<WalLogger>) -> Self {
-        Self { logger }
+        let first_frame_id_in_generation: u64 = logger.get_current_offset().try_into().unwrap();
+        Self {
+            logger,
+            replicas_with_hello: RwLock::new(HashSet::<SocketAddr>::new()),
+            hello_response: HelloResponse {
+                generation_id: uuid::Uuid::new_v4().to_string(),
```
[the remark this answers is not in the document]
Comment (MarinPostma) [reply sqld/src/rpc/wal_log.rs:60]: that would make it more consistent to how this is serialized in other protocols: https://github.com/libsql/sqld/blob/main/sqld/proto/proxy.proto#L6-L7

File: sqld/src/rpc/wal_log.rs [hunk 7]
```diff
@@ -74,8 +92,36 @@ impl WalLog for WalLogService {
...
         req: tonic::Request<LogOffset>,
     ) -> Result<tonic::Response<Self::LogEntriesStream>, Status> {
+        let replica_addr = req
+            .remote_addr()
+            .ok_or(Status::internal("No remote RPC address"))?;
+        {
+            let guard = self
+                .replicas_with_hello
+                .read()
+                .map_err(|_| Status::internal("Lock error"))?;
```
[the remark this answers is not in the document]
Comment (haaawk) [reply sqld/src/rpc/wal_log.rs:102]: fixed.

File: sqld/src/rpc/wal_log.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (haaawk) [reply sqld/src/rpc/wal_log.rs:60]: done

**Review by MarinPostma — APPROVED** [review]
bors merge

**Pull request merged**